docs/root/api-docs/xds_protocol.rst

```diff
@@ -91,18 +91,18 @@ combinations of two dimensions.
 
 The first dimension is State of the World (SotW) vs. incremental. The SotW approach was the
 original mechanism used by xDS, in which the client must specify all resource names it is
-interested in with each request (except when making a wildcard request in LDS/CDS), and the server
-must return all resources the client has subscribed to in each request (in LDS/CDS). This means
-that if the client is already subscribing to 99 resources and wants to add an additional one, it
-must send a request with all 100 resource names, rather than just the one new one. And the server
-must then respond by sending all 100 resources, even if the 99 that were already subscribed to have
-not changed (in LDS/CDS). This mechanism can be a scalability limitation, which is why the
-incremental protocol variant was introduced. The incremental approach allows both the client and
-server to indicate only deltas relative to their previous state -- i.e., the client can say that
-it wants to add or remove its subscription to a particular resource name without resending those
-that have not changed, and the server can send updates only for those resources that have changed.
-The incremental protocol also provides a mechanism for lazy loading of resources. For details on
-the incremental protocol, see :ref:`Incremental xDS <xds_protocol_delta>` below.
+interested in with each request, and the server must return all resources the client has
+subscribed to in each request (in LDS/CDS). This means that if the client is already subscribing
```

Comment:
I think `(in LDS/CDS)` part makes the sentence confusing. Perhaps it's worth clarifying that in LDS/CDS there could be multiple resources in a reply, while for RDS/EDS it's always one?

Reply:
To be clear, even in RDS and EDS, it's totally valid to have multiple resources per reply; for example, if the client is subscribed to 5 EDS resources and 2 of those resources change, the server can send a response including those two resources.  The thing that's special about LDS and CDS is that the server is required to send *all* resources that the client is subscribed to in every response, even the ones that haven't changed.  So if a client is subscribed to 5 LDS resources and 2 of them change, the server must send a response that includes all 5 resources.  (This is because the SotW protocol doesn't provide a way for the server to explicitly indicate to the client that a resource has been removed, so the only way to remove an LDS or CDS resource is by not including it in a response.  And the removal criteria for RDS and EDS resources is even messier; see the "Deleting Resources" section below.)

Anyway, I've attempted to reword this a bit to clarify it, but I haven't changed the phraseology about the condition this is discussing, because that was already correct.

Reply:
Thanks for the rewording, I think it's much clearer now. 

> To be clear, even in RDS and EDS, it's totally valid to have multiple resources per reply

Hmm, not according to the implementation (rds: https://github.com/envoyproxy/envoy/blob/main/source/common/router/rds_impl.cc#L227, eds: https://github.com/envoyproxy/envoy/blob/main/source/common/upstream/eds.cc#L166): updates with more than resource will be dropped.

Reply:
I believe that Envoy currently creates a separate stream for each resource when using non-aggregated EDS, as per #2943, and the same may be true of RDS.  However, this is a limitation of Envoy's implementation, not of the xDS protocol.  There is no protocol-level reason why there cannot be more than one resource send per response for any resource type.

Note that when using ADS, Envoy *does* support multiple resources per response for EDS and RDS.  But nothing about that functionality is actually tied to ADS; the exact same thing could be done in non-aggregated streams.